README.md
# Harvest [![Build Status](https://github.com/k1LoW/filt/workflows/build/badge.svg)](https://github.com/k1LoW/filt/actions) [![GitHub release](https://img.shields.io/github/release/k1LoW/harvest.svg)](https://github.com/k1LoW/harvest/releases) [![Go Report Card](https://goreportcard.com/badge/github.com/k1LoW/harvest)](https://goreportcard.com/report/github.com/k1LoW/harvest)

> Portable log aggregation tool for middle-scale system operation/troubleshooting.

![screencast](doc/screencast.svg)

Harvest provides the `hrv` command with the following features.

- Agentless.
- Portable.
- Only 1 config file.
- Fetch various remote/local log data via SSH/exec/Kubernetes API. ( `hrv fetch` )
- Output all fetched logs in the order of timestamp. ( `hrv cat` )
- Stream various remote/local logs via SSH/exec/Kubernetes API. ( `hrv stream` )
- Copy remote/local raw logs via SSH/exec. ( `hrv cp` )

## Quick Start ( for Kubernetes )

``` console
$ hrv generate-k8s-config > cluster.yml
$ hrv stream -c cluster.yml --tag='kube_apiserver or coredns' --with-path --with-timestamp
```

## Usage

### :beetle: Fetch and output remote/local log data

#### 1. Set log sources (and log type) in config.yml

``` yaml
---
targetSets:
  -
    description: webproxy syslog
    type: syslog
    sources:
      - 'ssh://webproxy.example.com/var/log/syslog*'
    tags:
      - webproxy
      - syslog
  -
    description: webproxy NGINX access log
    type: combinedLog
    sources:
      - 'ssh://webproxy.example.com/var/log/nginx/access_log*'
    tags:
      - webproxy
      - nginx
  -
    description: app log
    type: regexp
    regexp: 'time:([^\t]+)'
    timeFormat: 'Jan 02 15:04:05' # Golang time format and 'unixtime'
    timeZone: '+0900'
    sources:
      - 'ssh://app-1.example.com/var/log/ltsv.log*'
      - 'ssh://app-2.example.com/var/log/ltsv.log*'
      - 'ssh://app-3.example.com/var/log/ltsv.log*'
    tags:
      - app
  -
    description: db dump log
    type: regexp
    regexp: '"ts":"([^"]+)"'
    timeFormat: '2006-01-02T15:04:05.999-0700'
    sources:
      - 'ssh://db.example.com/var/log/tcpdp/eth0/dump*'
    tags:
      - db
      - query
  -
    description: PostgreSQL log
    type: regexp
    regexp: '^\[?(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} \w{3})'
    timeFormat: '2006-01-02 15:04:05 MST'
    multiLine: true
    sources:
      - 'ssh://db.example.com/var/log/postgresql/postgresql*'
    tags:
      - db
      - postgresql
  -
    description: local Apache access log
    type: combinedLog
    sources:
      - 'file:///path/to/httpd/access.log'
    tags:
      - httpd
-
    description: api on Kubernetes
    type: k8s
    sources:
      - 'k8s://context-name/namespace/pod-name*'
    tags:
      - api
      - k8s
```

You can use `hrv configtest` for config test.

``` console
$ hrv configtest -c config.yml
```

#### 2. Fetch target log data via SSH/exec/Kubernetes API ( `hrv fecth` )

``` console
$ hrv fetch -c config.yml --tag=webproxy,db
```

#### 3. Output log data ( `hrv cat` )

``` console
$ hrv cat harvest-20181215T2338+900.db --with-timestamp --with-host --with-path | less -R
```

#### 4. Count log data ( `hrv count` )

``` console
$ hrv count harvest-20191015T2338+900.db -g minute -g webproxy -b db
ts      webproxy db
2019-09-24 08:01:00     9618    5910
2019-09-24 08:02:00     9767    5672
2019-09-24 08:03:00     10815   7394
2019-09-24 08:04:00     11782   7109
2019-09-24 08:05:00     9896    6346
[...]
2019-09-24 08:24:00     11619   5646
2019-09-24 08:25:00     10541   6097
2019-09-24 08:26:00     11336   5264
2019-09-24 08:27:00     1102    5261
2019-09-24 08:28:00     1318    6660
2019-09-24 08:29:00     10362   5663
2019-09-24 08:30:00     11136   5373
2019-09-24 08:31:00     1748    1340
```

### :beetle: Stream remote/local logs

#### 1. [Set config.yml](#1-set-log-sources-and-log-type-in-configyml)

#### 2. Stream target logs via SSH/exec/Kubernetes API ( `hrv stream` )

``` console
$ hrv stream -c config.yml --with-timestamp --with-host --with-path --with-tag
```

### :beetle: Copy remote/local raw logs

#### 1. [Set config.yml](#1-set-log-sources-and-log-type-in-configyml)

#### 2. Copy remote/local raw logs to local directory via SSH/exec ( `hrv cp` )

``` console
$ hrv cp -c config.yml
```

### --tag filter operators

The following operators can be used to filter targets

`not`, `and`, `or`, `!`, `&&`, `||`

``` console
$ hrv stream -c config.yml --tag='webproxy or db' --with-timestamp --with-host --with-path
```

#### `,` is converted to ` or `

``` console
$ hrv stream -c config.yml --tag='webproxy,db'
```

is converted to

``` console
$ hrv stream -c config.yml --tag='webproxy or db'
```

### --source filter

filter targets using source regexp

``` console
$ hrv fetch -c config.yml --source='app-[0-9].example'
```

## Architecture

### `hrv fetch` and `hrv cat`

![img](doc/fetch.png)

### `hrv stream`

![img](doc/stream.png)

## Installation

```console
$ brew install k1LoW/tap/harvest
```

or

```console
$ go get github.com/k1LoW/harvest/cmd/hrv
```

## What is "middle-scale system"?

- < 50 instances
- < 1 million logs per `hrv fetch`

### What if you are operating a large-scale/super-large-scale/hyper-large-scale system?

Let's consider agent-base log collector/platform, service mesh and distributed tracing platform!

## Internal

- [harvest-*.db database schema](doc/schema)

## Requirements

- UNIX commands
  - date
  - find
  - grep
  - head
  - ls
  - tail
  - xargs
  - zcat
- sudo
- SQLite

## WANT

- tag DAG
- Viewer / Visualizer

## References

- [Hayabusa](https://github.com/hirolovesbeer/hayabusa): A Simple and Fast Full-Text Search Engine for Massive System Log Data
    - Make simple with a combination of commands.
    - Full-Text Search Engine using SQLite FTS.
- [stern](https://github.com/wercker/stern): ⎈ Multi pod and container log tailing for Kubernetes
    - Multiple Kubernetes log streaming architecture.


CHANGELOG.md
# Changelog

## [v0.17.3](https://github.com/k1LoW/harvest/compare/v0.17.2...v0.17.3) (2020-07-23)

* Update completion [#64](https://github.com/k1LoW/harvest/pull/64) ([k1LoW](https://github.com/k1LoW))

## [v0.17.2](https://github.com/k1LoW/harvest/compare/v0.17.1...v0.17.2) (2020-07-07)

* Fix log collect command for syslog (2 space like `Jun  6 19`) [#63](https://github.com/k1LoW/harvest/pull/63) ([k1LoW](https://github.com/k1LoW))

## [v0.17.1](https://github.com/k1LoW/harvest/compare/v0.17.0...v0.17.1) (2020-03-05)

* Fix zsh completion [#62](https://github.com/k1LoW/harvest/pull/62) ([k1LoW](https://github.com/k1LoW))

## [v0.17.0](https://github.com/k1LoW/harvest/compare/v0.16.2...v0.17.0) (2020-01-11)

* Add `hrv completion` [#61](https://github.com/k1LoW/harvest/pull/61) ([k1LoW](https://github.com/k1LoW))

## [v0.16.2](https://github.com/k1LoW/harvest/compare/v0.16.1...v0.16.2) (2019-11-27)

* Fix `hrv count` timestamp ordering [#60](https://github.com/k1LoW/harvest/pull/60) ([k1LoW](https://github.com/k1LoW))
* Use GitHub Actions [#59](https://github.com/k1LoW/harvest/pull/59) ([k1LoW](https://github.com/k1LoW))

## [v0.16.1](https://github.com/k1LoW/harvest/compare/v0.16.0...v0.16.1) (2019-10-18)

* Fix panic: runtime error: invalid memory address or nil pointer dereference [#58](https://github.com/k1LoW/harvest/pull/58) ([k1LoW](https://github.com/k1LoW))

## [v0.16.0](https://github.com/k1LoW/harvest/compare/v0.15.5...v0.16.0) (2019-10-18)

* Add `hrv info [DB_FILE]` [#57](https://github.com/k1LoW/harvest/pull/57) ([k1LoW](https://github.com/k1LoW))
* Fix timestamp grouping [#56](https://github.com/k1LoW/harvest/pull/56) ([k1LoW](https://github.com/k1LoW))
* Add table `metas` for saving `hrv fetch` info [#55](https://github.com/k1LoW/harvest/pull/55) ([k1LoW](https://github.com/k1LoW))

## [v0.15.5](https://github.com/k1LoW/harvest/compare/v0.15.4...v0.15.5) (2019-10-16)

* k8s client use --start-time and --end-time [#54](https://github.com/k1LoW/harvest/pull/54) ([k1LoW](https://github.com/k1LoW))
* Refactor RegexpParser [#53](https://github.com/k1LoW/harvest/pull/53) ([k1LoW](https://github.com/k1LoW))

## [v0.15.4](https://github.com/k1LoW/harvest/compare/v0.15.3...v0.15.4) (2019-10-15)

* Fix k8s timestamp parse [#52](https://github.com/k1LoW/harvest/pull/52) ([k1LoW](https://github.com/k1LoW))

## [v0.15.3](https://github.com/k1LoW/harvest/compare/v0.15.2...v0.15.3) (2019-10-10)

* SSH Client should fetch last line when SSH session closed. [#51](https://github.com/k1LoW/harvest/pull/51) ([k1LoW](https://github.com/k1LoW))
* Support parse 'unixtime' [#50](https://github.com/k1LoW/harvest/pull/50) ([k1LoW](https://github.com/k1LoW))

## [v0.15.2](https://github.com/k1LoW/harvest/compare/v0.15.1...v0.15.2) (2019-09-28)

* Fix: grep error Binary file (standard input) matches [#49](https://github.com/k1LoW/harvest/pull/49) ([k1LoW](https://github.com/k1LoW))

## [v0.15.1](https://github.com/k1LoW/harvest/compare/v0.15.0...v0.15.1) (2019-09-26)

* Fix configtest [#48](https://github.com/k1LoW/harvest/pull/48) ([k1LoW](https://github.com/k1LoW))

## [v0.15.0](https://github.com/k1LoW/harvest/compare/v0.14.2...v0.15.0) (2019-09-26)

* Add `hrv count` [#46](https://github.com/k1LoW/harvest/pull/46) ([k1LoW](https://github.com/k1LoW))

## [v0.14.2](https://github.com/k1LoW/harvest/compare/v0.14.1...v0.14.2) (2019-09-25)

* Fix fetch error handling [#47](https://github.com/k1LoW/harvest/pull/47) ([k1LoW](https://github.com/k1LoW))

## [v0.14.1](https://github.com/k1LoW/harvest/compare/v0.14.0...v0.14.1) (2019-09-21)

* Fix chan close timing (panic: send on closed channel) [#45](https://github.com/k1LoW/harvest/pull/45) ([k1LoW](https://github.com/k1LoW))

## [v0.14.0](https://github.com/k1LoW/harvest/compare/v0.13.1...v0.14.0) (2019-09-18)

* Do faster `hrv fetch` via ssh/file using grep timestamp [#44](https://github.com/k1LoW/harvest/pull/44) ([k1LoW](https://github.com/k1LoW))

## [v0.13.1](https://github.com/k1LoW/harvest/compare/v0.13.0...v0.13.1) (2019-09-17)

* Fix endtime (time.Time) is nil when no --end-time [#43](https://github.com/k1LoW/harvest/pull/43) ([k1LoW](https://github.com/k1LoW))

## [v0.13.0](https://github.com/k1LoW/harvest/compare/v0.12.0...v0.13.0) (2019-09-15)

* Add araddon/dataparse to parse `--*-time` option [#42](https://github.com/k1LoW/harvest/pull/42) ([k1LoW](https://github.com/k1LoW))
* Add `--duration` option [#41](https://github.com/k1LoW/harvest/pull/41) ([k1LoW](https://github.com/k1LoW))
* Fix `hrv cat` ts condition [#40](https://github.com/k1LoW/harvest/pull/40) ([k1LoW](https://github.com/k1LoW))
* Separate timestamp columns for grouping [#39](https://github.com/k1LoW/harvest/pull/39) ([k1LoW](https://github.com/k1LoW))
* Increase maxScanTokenSize [#38](https://github.com/k1LoW/harvest/pull/38) ([k1LoW](https://github.com/k1LoW))
* Refactor parser.Log struct [#37](https://github.com/k1LoW/harvest/pull/37) ([k1LoW](https://github.com/k1LoW))
* Refactor log pipeline [#36](https://github.com/k1LoW/harvest/pull/36) ([k1LoW](https://github.com/k1LoW))
* io.EOF is successful completion [#35](https://github.com/k1LoW/harvest/pull/35) ([k1LoW](https://github.com/k1LoW))

## [v0.12.0](https://github.com/k1LoW/harvest/compare/v0.11.0...v0.12.0) (2019-08-19)

* Use STDERR instead of STDOUT [#34](https://github.com/k1LoW/harvest/pull/34) ([k1LoW](https://github.com/k1LoW))
* Add gosec [#33](https://github.com/k1LoW/harvest/pull/33) ([k1LoW](https://github.com/k1LoW))
* Add --verbose option [#32](https://github.com/k1LoW/harvest/pull/32) ([k1LoW](https://github.com/k1LoW))

## [v0.11.0](https://github.com/k1LoW/harvest/compare/v0.10.0...v0.11.0) (2019-06-24)

* Add `hrv generate-k8s-config` [#31](https://github.com/k1LoW/harvest/pull/31) ([k1LoW](https://github.com/k1LoW))

## [v0.10.0](https://github.com/k1LoW/harvest/compare/v0.9.0...v0.10.0) (2019-06-23)


cmd/cat.go
// Copyright © 2019 Ken'ichiro Oyama <k1lowxb@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

package cmd

import (
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/k1LoW/harvest/db"
	"github.com/k1LoW/harvest/logger"
	"github.com/k1LoW/harvest/stdout"
	"github.com/labstack/gommon/color"
	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	"go.uber.org/zap"
)

const (
	delimiter = ","
)

var (
	match string
	st    string
	et    string
)

// catCmd represents the cat command
var catCmd = &cobra.Command{
	Use:   "cat [DB_FILE]",
	Short: "cat",
	Long:  `cat`,
	Args: func(cmd *cobra.Command, args []string) error {
		if len(args) != 1 {
			return errors.WithStack(errors.New("requires one arg"))
		}
		return nil
	},
	Run: func(cmd *cobra.Command, args []string) {
		l := logger.NewLogger()
		dbPath := args[0]

		if _, err := os.Lstat(dbPath); err != nil {
			l.Error(fmt.Sprintf("%s not exists", dbPath), zap.String("error", err.Error()))
			os.Exit(1)
		}

		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		d, err := db.AttachDB(ctx, l, dbPath)
		if err != nil {
			l.Error("DB attach error", zap.String("error", err.Error()))
			os.Exit(1)
		}

		cond, err := buildCondition()
		if err != nil {
			l.Error("option error", zap.String("error", err.Error()))
			os.Exit(1)
		}

		hLen, tLen, err := getCatStdoutLengthes(d, withHost, withPath, withTag)
		if err != nil {
			l.Error("option error", zap.String("error", err.Error()))
			os.Exit(1)
		}

		hosts, err := d.GetHosts()
		if err != nil {
			l.Error("DB query error", zap.String("error", err.Error()))
			os.Exit(1)
		}

		sout, err := stdout.NewStdout(
			withTimestamp,
			withTimestampNano,
			withHost,
			withPath,
			withTag,
			hLen,
			tLen,
			noColor,
		)
		if err != nil {
			l.Error("fetch error", zap.String("error", err.Error()))
			os.Exit(1)
		}

		err = sout.Out(d.Cat(cond), hosts)
		if err != nil {
			l.Error("fetch error", zap.String("error", err.Error()))
			os.Exit(1)
		}
	},
}

func colorizeTag(colorFunc func(interface{}, ...string) string, tag string) string {
	colorized := []string{}
	tags := strings.Split(tag, " ")
	for _, t := range tags {
		colorized = append(colorized, colorFunc(t, color.B))
	}
	return strings.Join(colorized, " ")
}

// buildCondition ...
func buildCondition() (string, error) {
	matchCond := []string{}
	cond := []string{}
	if match != "" {
		matchCond = append(matchCond, match)
	}
	loc, err := time.LoadLocation("Local")
	if err != nil {
		return "", err
	}
	if st != "" {
		t, err := time.ParseInLocation("2006-01-02 15:04:05", st, loc)
		if err != nil {
			return "", err
		}
		cond = append(cond, fmt.Sprintf("ts >= %d", t.UnixNano()))
	}
	if et != "" {
		t, err := time.ParseInLocation("2006-01-02 15:04:05", et, loc)
		if err != nil {
			return "", err
		}
		cond = append(cond, fmt.Sprintf("ts <= %d", t.UnixNano()))
	}
	if tag != "" {
		cond = append(cond, fmt.Sprintf("( tag LIKE '%%[%s]%%' )", strings.Join(strings.Split(tag, delimiter), "]%' OR tag LIKE '%[")))
	}

	if len(matchCond) > 0 {
		cond = append(cond, fmt.Sprintf("content MATCH '%s'", strings.Join(matchCond, " AND ")))
	}

	if len(cond) == 0 {
		return "", nil
	}

	return fmt.Sprintf(" WHERE %s", strings.Join(cond, " AND ")), nil
}

func getCatStdoutLengthes(d *db.DB, withHost, withPath, withTag bool) (int, int, error) {
	var (
		hLen int
		tLen int
		err  error
	)
	if withHost && withPath {
		hLen, err = d.GetColumnMaxLength("host", "path")
		if err != nil {
			return 0, 0, err
		}
	} else if withHost {
		hLen, err = d.GetColumnMaxLength("host")
		if err != nil {
			return 0, 0, err
		}
	} else if withPath {
		hLen, err = d.GetColumnMaxLength("path")
		if err != nil {
			return 0, 0, err
		}
	}
	if withTag {
		tLen, err = d.GetColumnMaxLength("tag")
		if err != nil {
			return 0, 0, err
		}
	}
	return hLen, tLen, nil
}

func init() {
	rootCmd.AddCommand(catCmd)
	catCmd.Flags().BoolVarP(&withTimestamp, "with-timestamp", "", false, "output with timestamp")
	catCmd.Flags().BoolVarP(&withTimestampNano, "with-timestamp-nano", "", false, "output with timestamp nano sec")
	catCmd.Flags().BoolVarP(&withHost, "with-host", "", false, "output with host")
	catCmd.Flags().BoolVarP(&withPath, "with-path", "", false, "output with path")
	catCmd.Flags().BoolVarP(&withTag, "with-tag", "", false, "output with tag")
	catCmd.Flags().StringVarP(&match, "match", "", "", "filter logs using SQLite FTS `MATCH` query")
	catCmd.Flags().StringVarP(&tag, "tag", "", "", "filter logs using tag (format: foo,bar)")
	catCmd.Flags().StringVarP(&st, "start-time", "", "", "log start time (format: 2006-01-02 15:04:05)")
	catCmd.Flags().StringVarP(&et, "end-time", "", "", "log end time (format: 2006-01-02 15:04:05)")
	catCmd.Flags().BoolVarP(&noColor, "no-color", "", false, "disable colorize output")
}
